### DISCUSSION TITLE
Dynamic Handling of nested JSONs

### DISCUSSION THREAD

**Discussion by GurramLakshmiNarayana** [nlohmann/json#3568] [Q&A] [answered]
Iam trying to add a JSON into a existing JSON Dynamically(i.e the nested JSON is not initialized at startup but as the time progresses, i will be adding a json in to the main json).
For example lets say my initial config is :

{
    "version": "1.0.0",
    "boardtype": "1"
}

Then i want to add a JSON in side this main json like this:

{
    "version": "1.0.0",
    "boardtype": "1",
    "pointerOffset": {
        "X": 0.12,
        "Y": -1.8,
        "Z": 3.2
    },
    "platformOffset": {
        "X": 0.12,
        "Y": -1.8,
        "Z": 3.2
    }
}

From the example you can see that the JSON file that iam trying to create is a device config file and it can change drastically over time.
My requirement is as follows:
1. I should be able to ADD, DELETE and MODIFY nested JSONs dynamically without static declarations
2. The paths and the names of the keys are also dynamic, unknown and the APIs should handle no matter the PATH

I have been looking at various tutorials and i was mostly able to find may be how to modify the contents using PATCH and MERGE and also mostly delete but was unable to find a way on how to add a JSON dynamically inside a existing JSON.
Can you tell me if this is possible with this library and how to do it with simple examples.
Note: The neither names(i.e of the path), nor the creation/Addition of nested JSONs should have static declarations

**Comment by falbrechtskirchinger** [comment]
Is this what you're looking for? https://json.nlohmann.me/api/basic_json/operator%5B%5D/

Otherwise, please try again, preferably using code instead of plain English.

**Comment by GurramLakshmiNarayana** [comment]
No that is not what iam looking for
As i mentioned in the description, the depth of nesting is unknown and even the names of the objects is undefined and comes in dynamically or gets deleted dynamically.
When it comes to code, i don't have any implementation yet as iam doing a feasibility study of the JSON libraries for my requirement.
So far i found : [https://github.com/json-c/json-c](url) which matches my requirement but is in c and requires lot of code to modify contents of JSON.
The reason why iam trying to use this library is because it is in C++ and it is just like static linking just by including a header file

**Reply by falbrechtskirchinger** [reply]
I'll refer you back to my first answer. I think `operator[]` and related functions are exactly what you want. It accepts a dynamic key in form of a `std::string`, `const char *`, `std::string_view`, etc. Just repeat the process to get a reference to a deeply nested element. Alternatively, you can use a [`json_pointer`](https://json.nlohmann.me/api/json_pointer/) to obtain a reference to a deeply nested element in one go.

**Reply by GurramLakshmiNarayana** [reply]
Buddy i know how to access elements in nested JSON using JSON pointer or PATCH/MERGE methods.
Iam not speaking about accessing nested JSONs dynamically
Iam discussing about creating nested JSONs dynamically
Can you please be specific in your answers

**Reply by falbrechtskirchinger** [reply]
Accessing and creating are pretty much the same thing. Accessing a key via `operator[]` creates a `null` JSON value. See my other answer for how to create a JSON object explicitly.

**Comment by GurramLakshmiNarayana** [comment]
To make it clear I just want a example on how to add a json(even a empty one) in to a existing JSON:

`json Jmain = R"({"version": "1.0.0", "boardtype": 1})"_json;`

then i want to do thsi:
```
std::string subKey = "/pointerOffset";
json::json_pointer jptr(subKey);
```

ten do some thing like :
`Jmain[jptr] = <new json object(or  some thing similar)>`

which will make my JSON:
{
"version": "1.0.0",
"boardtype": "1",
"pointerOffset": {}
}

**Reply by falbrechtskirchinger** [reply]
Thanks. You should have opened with that. :-)

**Comment by falbrechtskirchinger** [comment, marked as answer]
You pretty much had it.

```cpp
json Jmain = R"({"version": "1.0.0", "boardtype": 1})"_json;

std::string subKey = "pointerOffset";
Jmain[subKey] = json::object(); // This step is technically not required; accessing a `null` value will convert it to an object automatically

json::json_pointer jptr{"/pointerOffset/X"};
Jmain[jptr] = 0.12;
```

Which will give you:
```json
{
  "version": "1.0.0",
  "boardtype": "1",
  "pointerOffset": {
    "X": 0.12
  }
}
```

Edit: See the notes section here: https://json.nlohmann.me/api/basic_json/operator%5B%5D/#notes

> If key is not found in the object, then it is silently added to the object and filled with a null value to make key a valid reference. In case the value was null before, it is converted to an object.

**Reply by GurramLakshmiNarayana** [reply]
Thank You very much for clearing the confusion and great explanation

**Comment by jwalkerbg** [comment]
Hi,

May be you want to use json::emplace
(https://json.nlohmann.me/api/basic_json/emplace/)

First you have
json root = R"_({
"version": "1.0.0",
"boardtype": "1"
})_"_json;

Then, later you create new json objects dynaically.
Then you can emplace these objects:

std::string key = CreateKey();
json dynamic = CreateIt();

root.emplace(key,dynamic);


May this will give you an idea for how to do it.

Regards,
Ivan Cenov
На 01.7.2022 г. в 20:51, GurramLakshmiNarayana написа:
>
> Iam trying to add a JSON into a existing JSON Dynamically(i.e the 
> nested JSON is not initialized at startup but as the time progresses, 
> i will be adding a json in to the main json).
> For example lets say my initial config is :
>
> {
> "version": "1.0.0",
> "boardtype": "1"
> }
>
> Then i want to add a JSON in side this main json like this:
>
> {
> "version": "1.0.0",
> "boardtype": "1",
> "pointerOffset": {
> "X": 0.12,
> "Y": -1.8,
> "Z": 3.2
> },
> "platformOffset": {
> "X": 0.12,
> "Y": -1.8,
> "Z": 3.2
> }
> }
>
> From the example you can see that the JSON file that iam trying to 
> create is a device config file and it can change drastically over time.
> My requirement is as follows:
>
>  1. I should be able to ADD, DELETE and MODIFY nested JSONs
>     dynamically without static declarations
>  2. The paths and the names of the keys are also dynamic, unknown and
>     the APIs should handle no matter the PATH
>
> I have been looking at various tutorials and i was mostly able to find 
> may be how to modify the contents using PATCH and MERGE and also 
> mostly delete but was unable to find a way on how to add a JSON 
> dynamically inside a existing JSON.
> Can you tell me if this is possible with this library and how to do it 
> with simple examples.
> Note: The neither names(i.e of the path), nor the creation/Addition of 
> nested JSONs should have static declarations
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/nlohmann/json/discussions/3568>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAD4UQSF753BSSQMDTD2L5TVR4V25ANCNFSM52NNMKGQ>.
> You are receiving this because you are subscribed to this 
> thread.Message ID: ***@***.***>
>
-- 
Regards,

Ivan Cenov
Software Manager
OKTO-7 Ltd, Botevgrad, Bulgaria
***@***.******@***.***
Redmine - Bulgarian Translation Maintainer

=== Show more empathy in these COVID & economic crisis times ===
